### ISSUE TITLE
Streams freezes above 20 viewers

### ISSUE THREAD

**Issue by nangavar** [AirenSoft/OvenMediaEngine#149]
Hi All!

Im using OME with about 8 different streams using RTPS Pulling provider.

When a single stream has more than 20 viewers, this stream starts freezing in all viewers, but the other streams are not affected.

Networking seems to be OK (have the same behaviour in cloud servers), max priority on firewalls. 
Bandwidth seems to be OK (other streams are not affected)
Hardware resources seems to be ok (Intel Xeon 64cores, 256GB RAM, 2% CPU usage)

Below my Server.xml file.

Any idea? I couldn't find any problem at network interfaces, or firewalls logs.

Am I missing something? is there a special config for network interfaces in Linux that I have to perform?
```xml
<?xml version="1.0" encoding="UTF-8" ?>

<Server version="7">
	<Name>OvenMediaEngine</Name>
	<!-- Host type (origin/edge) -->
	<Type>origin</Type>
	<!-- Specify IP address to bind (* means all IPs) -->
	<IP>*</IP>
	<!-- Settings for the ports to bind -->
	<Bind>
		<Providers>
			<RTMP>
				<Port>1935</Port>
			</RTMP>
		</Providers>

		<Publishers>
			<OVT>
				<Port>9000</Port>
			</OVT>
			<HLS>
				<!--<Port>80</Port>-->
				<!-- If you want to use TLS, specify the TLS port -->
				<!-- <TlsPort>443</TlsPort> -->
			</HLS>
			<DASH>
				<!-- <Port>80</Port>-->
				<!-- If you want to use TLS, specify the TLS port -->
				<!-- <TlsPort>443</TlsPort> -->
			</DASH>
		<WebRTC>
				<Signalling>
					<Port>3333</Port>
					<!-- If you want to use TLS, specify the TLS port -->
					<!-- <TlsPort>3334</TlsPort> -->
				</Signalling>
				<IceCandidates>
					<IceCandidate>*:10000-10005/udp</IceCandidate>
				</IceCandidates>
			</WebRTC>
		</Publishers>
	</Bind>

	<!-- P2P works only in WebRTC -->
	<!--
	<P2P>
		<MaxClientPeersPerHostPeer>2</MaxClientPeersPerHostPeer>
	</P2P>
	-->

	<VirtualHosts>
		<!-- You can use wildcard like this to include multiple XMLs -->
		<VirtualHost include="VHost*.xml" />
		<VirtualHost>
			<Name>default</Name>

			<!-- Settings for multi domain and TLS -->
			<Domain>
				<Names>
					<!--
						Domain names
						<Name>stream1.airensoft.com</Name>
						<Name>stream2.airensoft.com</Name>
						<Name>*.sub.airensoft.com</Name>
					-->
					<Name>*</Name>
				</Names>
				<!--
				<TLS>
					<CertPath>path/to/file.crt</CertPath>
					<KeyPath>path/to/file.key</KeyPath>
					<ChainCertPath>path/to/file.crt</ChainCertPath>
				</TLS>
				-->
			</Domain>
			
			<Origins>
				<Origin>
					<Location>/app2/stream1</Location>
					<Pass>
						<Scheme>rtsp</Scheme>
						<Urls><Url>encoder_IP/streamkey</Url></Urls>
					</Pass>
				</Origin>

				<Origin>
					<Location>/app2/stream2</Location>
					<Pass>
						<Scheme>rtsp</Scheme>
						<Urls><Url>encoder_IP/streamkey</Url></Urls>
					</Pass>
				</Origin>
				<Origin>
					<Location>/app2/stream3</Location>
					<Pass>
						<Scheme>rtsp</Scheme>
						<Urls><Url>encoder_IP/streamkey</Url></Urls>
					</Pass>
				</Origin>
				<Origin>
					<Location>/app2/stream4</Location>
					<Pass>
						<Scheme>rtsp</Scheme>
						<Urls><Url>encoder_IP/streamkey</Url></Urls>
					</Pass>
				</Origin>
				<Origin>
					<Location>/app2/stream5</Location>
					<Pass>
						<Scheme>rtsp</Scheme>
						<Urls><Url>encoder_IP/streamkey</Url></Urls>
					</Pass>
				</Origin>
				<Origin>
					<Location>/app2/stream6</Location>
					<Pass>
						<Scheme>rtsp</Scheme>
						<Urls><Url>encoder_IP/streamkey</Url></Urls>
					</Pass>
				</Origin>
				<Origin>
					<Location>/app2/stream7</Location>
					<Pass>
						<Scheme>rtsp</Scheme>
						<Urls><Url>encoder_IP/streamkey</Url></Urls>
					</Pass>
				</Origin>
				<Origin>
					<Location>/app2/stream8</Location>
					<Pass>
						<Scheme>rtsp</Scheme>
						<Urls><Url>encoder_IP/streamkey</Url></Urls>
					</Pass>
				</Origin>
			</Origins>
		
			<!-- Settings for application RTMP -->
			<Applications>
				<Application>
					<Name>app</Name>
					<Type>live</Type>
					
					<Encodes>
						<Encode>
							<Name>opus</Name>
							<Audio>
								<Codec>opus</Codec>
								<Bitrate>128000</Bitrate>
								<Samplerate>48000</Samplerate>
								<Channel>2</Channel>
							</Audio>
							<Video>
								<Bypass>true</Bypass>
							</Video>
						</Encode>
					</Encodes>
					
					<Streams>
						<Stream>
							<Name>${OriginStreamName}</Name>
							<Profiles>
								<Profile>opus</Profile> 
							</Profiles>
						</Stream>
					</Streams>
					
					<Providers>
						<RTMP />
					</Providers>
					
					<Publishers>
						<ThreadCount>64</ThreadCount>
						<WebRTC>
							<Timeout>30000</Timeout>
						</WebRTC>					
					</Publishers>	
				</Application>
			<!-- Settings for application RTSP Pulling -->			
				<Application>
					<Name>app2</Name>
					<Type>live</Type>
					
					<Encodes>
						<Encode>
							<Name>opus</Name>
							<Audio>
								<Codec>opus</Codec>
								<Bitrate>128000</Bitrate>
								<Samplerate>48000</Samplerate>
								<Channel>2</Channel>
							</Audio>
							<Video>
								<Bypass>true</Bypass>
							</Video>
						</Encode>
					</Encodes>
					
					<Streams>
						<Stream>
							<Name>${OriginStreamName}</Name>
							<Profiles>
								<Profile>opus</Profile> 
							</Profiles>
						</Stream>
					</Streams>
					
					<Providers>
						<RTSPPull />
					</Providers>
					
					<Publishers>
						<ThreadCount>128</ThreadCount>
						<WebRTC>
							<Timeout>30000</Timeout>
						</WebRTC>					
					</Publishers>
				</Application>
				
			</Applications>
		</VirtualHost>
	</VirtualHosts>
</Server>
```

**Comment by getroot** [comment]
How about playing the same stream in HLS? If HLS is normal, there may be a high UDP packet loss in your network. What video bitrates are input?
 
Or this may be highly related to the kernel network configuration.

Alternatively, the ThreadCount of Publishers is set to 128, which may be too high. I think 16 is enough. (This is not a good idea because it is the number of threads created per stream.)

We have experience of running 1500 RTSP -> 1500 WebRTC/HLS, 1 RTSP -> 2000 HLS on a 40 Core CPU, 10Gbps network, but it is not exactly the same as your configuration. We'll take a look at your settings.

**Comment by nangavar** [comment]
We are using only WebRTC in order to get the lowest possible latency. We are using OME for live TV production, so latency is mandatory. But I can try with HLS in order to address the problem. We are using 2Mbps streams.

Regarding kernel network configuration, what should I consider?

I have changed the threads to 16. 128 was an attempt to improve the stream performance. I was using 32 actually.

It is weird because the others streams running are not affected, only the stream with more than 20 viewers is affected. Will keep trying and let you know.

**Comment by getroot** [comment]
Did you... Did you run 20 WebRTC Players on one PC? Chrome has a limit on the maximum playback channel.

**Comment by nangavar** [comment]
Nope, with different viewers. One by browser.

I tried with three different hardwares (cloud and physical), three different ISPs, with and without Firewall. Always same, above 20 users stream starts to get freezes. I don't get any performance message on linux, Im using NETDATA to monitor.

**Comment by getroot** [comment]
Enter chrome://webrtc-internals in the address bar of the CHROME browser and check the packet loss. However, considering that other streams are working, this is unlikely to be the cause.

The following article will help you with kernel tuning. However, we usually do kernel tuning for over 500 viewers. No kernel tuning is actually required for 20 viewers... https://liujarry.wordpress.com/2016/10/12/tcp-and-udp-tuning-in-linux/

It's really weird. If it is 200, it may be depending on the kernel setting. But... 20?

I am going to test if something is wrong with the latest version.

(It may not be, but if you have multiple one by browsers on a single PC, it is also limited by Chrome. I think it is wrong to write one by device.)

**Comment by getroot** [comment]
Also, please test whether the freeze stream works well when played in HLS.

**Comment by nangavar** [comment]
Will try and let you know of course. It will take me a while to arrange the test with viewers. But i will come back with results.

**Comment by nangavar** [comment]
We ran the test with different viewers at their own homes. I mean one player by device.

**Comment by nangavar** [comment]
After some tests I found a "UDP send buffer error"
I have performed the kernel tuning but I get same behaviour. Attached some images of the buffer errors.
These errors start above the viewer number 19. Below 19 no errors. I tried with at least three servers cloud-based and on-premise, different ISPs. Always same behaviour.
Any idea? 

![Screen Shot 2020-07-15 at 21 14 22](https://user-images.githubusercontent.com/66489890/87612166-2ee5af80-c6e0-11ea-8085-e3f5f059496d.png)

![Screen Shot 2020-07-15 at 20 57 20](https://user-images.githubusercontent.com/66489890/87612244-618fa800-c6e0-11ea-8d51-750941962f10.png)
![Screen Shot 2020-07-15 at 20 57 36](https://user-images.githubusercontent.com/66489890/87612247-63f20200-c6e0-11ea-9ddb-48e3399c3b36.png)

**Comment by getroot** [comment]
UDP send buffer errors are the cause of the problem. In this case, HLS will work fine. Because it is TCP. (And TCP has a retransmission capability.) Your test was great. Now you can solve it.

Here are a few things I know:

1. Check the send buffer size of the kernel with the following command.

sysctl net.core.wmem_max
sysctl net.core.wmem_default

This value should be large enough. Otherwise, it will not keep up with the speed of the 16 threads you set.

sysctl -w net.core.wmem_default = 4194304
sysctl -w net.core.wmem_max = 4194304

2. What is the speed of your NIC? Is it 1G or 10G? Check if it is set to 100Mbps.

**Comment by nangavar** [comment]
Hi getroot!
No more buffer errors :) (0 errors) However streams still getting freeze but now above 25~30 streams 
Im waiting my cloud-server provider to open 8080 port to run the HLS test. I will let you know the results.

What I did to stop getting buffering errors:

- I added on "/etc/sysctl.conf":
net.ipv4.udp_rmem_min=8388608
net.ipv4.udp_wmem_min=8388608
net.core.netdev_max_backlog=8388608
net.core.wmem_default = 8388608
net.core.wmem_max = 8388608
net.core.rmem_max = 8388608

- I extended the TxqueueLen:
ifconfig <NIC> txqueuelen 10000

**Comment by getroot** [comment]
We tested it by installing the latest version of OME on our server once again. We have confirmed that 70 clients are playing well. It would be a good idea to check your server's UDP network performance with iperf.

**Comment by nangavar** [comment]
We finally performed a load test, and the behaviour improves a lot! Now im able to stream above 40 streams without drops and buffer errors.
From 40 and above I start receiving the exceeded queue message. I have 16 threads per stream configured. Should I increase this number?:

My logs:
```log
[2020-07-17 13:07:24.873] W 5584 ov.Queue | queue.h:179  | [0x7ff332e01c58] WebRTC Publisher Application/#default#app2/stream4 StreamWorker Queue size has exceeded the threshold: queue: 100, threshold: 100
[2020-07-17 13:07:25.871] W 5584 ov.Queue | queue.h:179  | [0x7ff332e01ed8] WebRTC Publisher Application/#default#app2/stream4 StreamWorker Queue size has exceeded the threshold: queue: 100, threshold: 100
[2020-07-17 13:07:26.877] W 5584 ov.Queue | queue.h:179  | [0x7ff332e023d8] WebRTC Publisher Application/#default#app2/stream4 StreamWorker Queue size has exceeded the threshold: queue: 100, threshold: 100
```

**Comment by nangavar** [comment]
Finally I get 65 streams without freezes, I had to increase ThreadCount and no more queue messages!

For the future, if I wanna serve around 500 clients, what should I change?

I will close this issue! 
Many thanks to you guys! :)

**Comment by getroot** [comment]
Queue messages are not really a problem with streaming. This log is used for reference to find a suitable threshold. If it continues to increase at 100 it may be a problem, but if it continues to be constant it does not matter. I will modify the threshold of log messages.

What solved your problem? (Kernel socket settings?) Please share the cause and solution of your problem.

**Comment by nangavar** [comment]
Hi!
What I did was kernel tuning and some changes in my cloud-server provider.

1)CLOUD-SERVER PROVIDER
They had to "move" the server that was behind a virtualised firewall to the main one. We were experiencing some drops in the service since we were behind a VMWare FireWall.
However after moving the server we still getting the udp_buffer error. So we did kernel tuning.

2)KERNEL TUNNING
I added on "/etc/sysctl.conf":
- net.ipv4.udp_rmem_min=8388608
net.ipv4.udp_wmem_min=8388608
net.core.netdev_max_backlog=8388608
net.core.wmem_default = 8388608
net.core.wmem_max = 8388608
net.core.rmem_max = 8388608

I extended the TxqueueLen:
- ifconfig NIC txqueuelen 10000

Hope it helps if anybody find same issue.
Best
nangavar
